### PR TITLE
Migration to rocksdb-haskell-ng

### PR DESCRIPTION
**Pull request by chrisdone**
Summary:

* Comments out a few RocksDB-related things that aren't currently used.
* Switches all cabal-depends to rocksdb-haskell-ng.
* Adds rocksdb-haskell-ng as an extra-dep to the stack.yaml.

The rocksdb-haskell-ng project is a ground-up rewrite (with some code re-used where possible) of the rocksdb-haskell project, with things that are not used by cardano-sl not included.

The rewrite is motivated by making the binding as safe as possible: it should not ever segfault, and should be safe from memory leaks, double closes, async exceptions, and safe concurrent use, all for both connection handles and for iterators. The worst case scenario is that it should throw a meaningful exception when you have made a mistake.

To backup these aims, included is a fairly comprehensive test suite which attempts to test edge-cases and programmer error. Included are tests made by Niklas and some fixes for Unicode file access on Linux and Windows. It also includes a benchmark suite to track any performance improvements.

It also has Linux and Windows CI.

[Linux build Status](https://travis-ci.org/chrisdone/rocksdb-haskell-ng) [Windows build status](https://ci.appveyor.com/project/chrisdone/rocksdb-haskell-ng)

### CODE REVIEW

**Comment by domenkozar** [comment]
You'll have to run `pkgs/generate.sh`

**Comment by chrisdone** [comment]
I managed to install nix and it's currently running.  How long does it normally take?

**Comment by domenkozar** [comment]
~1min

**Comment by chrisdone** [comment]
Took 6mins on my machine (probably because it's a first run). At the end it failed with:

``` 
downloading ‘https://github.com/luite/hfsevents.git’... [1D [0/0 KiB, 0.0 KiB/s][21D [0/0 KiB, 0.0 KiB/s]
error: unable to download ‘https://github.com/luite/hfsevents.git’: Problem with the SSL CA cert (path? access rights?) (77)
Initialized empty Git repository in /private/var/folders/2s/6j1qqrgd57bg9ctkltrkn3z80000gn/T/git-checkout-tmp-iogI7e91/hfsevents-25a53d4/.git/
fatal: unable to access 'https://github.com/luite/hfsevents.git/': error setting certificate verify locations:
  CAfile: /etc/ssl/cert.pem
  CApath: none
fatal: unable to access 'https://github.com/luite/hfsevents.git/': error setting certificate verify locations:
  CAfile: /etc/ssl/cert.pem
  CApath: none
Unable to checkout 25a53d417d7c7a8fc3116b63e3ba14ca7c8f188f from https://github.com/luite/hfsevents.git.
abort: HTTP Error 406: Not Acceptable
svn: E205000: Syntax error in revision argument '25a53d417d7c7a8fc3116b63e3ba14ca7c8f188f'
failed to open trace file: [Errno 2] No such file or directory: '/homeless-shelter/.bzr.log'
Not checking SSL certificate for github.com.
bzr: ERROR: Not a branch: "https://github.com/luite/hfsevents.git/".
cabal2nix: user error (Failed to fetch source. Does the URL exist?)
```

Could I maybe give you push rights to the repo and you run it and push the result to this PR?

**Comment by neongreen** [comment]
You can usually just steal the diff from Travis and apply it – no need to have Nix installed locally.

**Comment by domenkozar** [comment]
This is because git doesn't find CA certs. Could you try applying:

```
diff --git a/pkgs/generate.sh b/pkgs/generate.sh
index 126ba9ece..a25e07b31 100755
--- a/pkgs/generate.sh
+++ b/pkgs/generate.sh
@@ -4,7 +4,7 @@
 # regenerate the `pkgs/default.nix` file based on the current contents of cardano-sl.cabal and stack.yaml
 
 function runInShell {
-  nix-shell -j 4 -p nix cabal2nix glibcLocales --pure --run "LANG=en_US.utf-8 NIX_REMOTE=$NIX_REMOTE NIX_PATH=$NIX_PATH $*"
+  nix-shell -j 4 -p git curl nix cabal2nix glibcLocales --pure --run "LANG=en_US.utf-8 NIX_REMOTE=$NIX_REMOTE NIX_PATH=$NIX_PATH $*"
 }
 
 set -xe
```

and re-run `pkgs/generate.sh`?

**Comment by chrisdone** [comment]
@domenkozar Sadly that change didn't change the result.

**Comment by rockbmb** [comment]
@chrisdone where can I find the Unicode fix for Windows? I'd like to know that, if possible.

**Comment by chrisdone** [comment]
@rockbmb The unicode test doesn't seem to fail on Windows. But that doesn't indicate there's no problem on Windows. Now that you mention it, I believe it's worth adding a Windows-specific test. 

If you have a Windows test case that we can add as a regression test, that would be handy. Otherwise I'll have to construct one.

**Comment by rockbmb** [comment]
@chrisdone The reason I ask is that I wrote that Unicode test, and it does pass on Windows; however, the problem persists.

I am still trying to uncover the reasons why, I'll use this PR's "Conversation" section to keep you updated on what I find.

**Comment by domenkozar** [comment]
cc @adinapoli-iohk to advise how soon we'll need rocksdb snapshots.

**Comment by domenkozar** [comment]
Superseeded by https://github.com/input-output-hk/cardano-sl/pull/2368 to get CI rolling.
